### ISSUE TITLE
How to use this dataset?

### ISSUE THREAD

**Issue by moryekram** [manitadayon/Auto_HMM#1]
Hi Manitadayon,
I am glad to find your code and thanks from sharig it.
I am new in machine learning. Now I want to code for [this paper](https://www.researchgate.net/profile/Daniel-Nieto-Mora/publication/349725411_Using_Hidden_Markov_Models_for_Profiling_Driver_Behavior_Patterns/links/603ec8444585154e8c7245dd/Using-Hidden-Markov-Models-for-Profiling-Driver-Behavior-Patterns.pdf) and want to use [this dataset ](https://github.com/jair-jr/driverBehaviorDataset). But i don't know how can i do this. Can you help me?
I don't know exactly what is states in this paper and how give dataset to your code, and how wite code for this paper.

**Comment by manitadayon** [comment]
Hi For sure. I have a video on my YouTube channel on how to model dataset using a hidden Markov model : Hidden Markov Model (Modeling and Implementation) 

https://www.youtube.com/watch?v=1b-sd7gulFk&t=4s&ab_channel=AIandMLFundamentals

  
Please let me know if you still have questions. 
Thanks
Manie 


Sent from Yahoo Mail for iPhone


On Friday, January 21, 2022, 4:10 AM, moryekram ***@***.***> wrote:




Hi Manitadayon,
I am glad to find your code and thanks from sharig it.
No i want to code for this paper and want to use this dataset . But i don't know how can i do this. Can you help me?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by moryekram** [comment]
Hello again. I watched your video but I don't know where to start. In
paper, the first exec for 1,3,5,7,9 states and said that for example 5
states are better. But I don't know how to pass my dataset to your code ?
How to get accuracy of multiple states? and  so on. And excuse me, are you
iranian? Can I write in Persian language? O may I have chat in social
networks(because I have to deliver my project this monday)

**Comment by manitadayon** [comment]
Hi again, I posted another video just now and I explain Auto-HMM in detail.

**Comment by moryekram** [comment]
Unfortunately gives error "more than 23259samples in lengths array" when using Hmm_Testing. This is my dataset
[giroscopio_terra-1.csv](https://github.com/manitadayon/Auto_HMM/files/7918930/giroscopio_terra-1.csv)
this file is merge of 4 file with deletion rows that are not in any class
, and this is my parameters
`Train_ratio=0.7
Cov_Type='diag'
Max_state=3
Max_mixture=4
Iter=100
Feat=3
N=11077
T=69
flag=1
col_list = ["uptimeNanos", "x", "y", "z"]
Path='mydataset/16/giroscopio_terra-1.csv'
Data=pd.read_csv(Path, header=0, parse_dates=[0], index_col=0, usecols=col_list)
Exam_HMM=Supervised_HMM(Train_ratio,Cov_Type,Max_state,Max_mixture,Iter,Feat,N,T,Data,flag)
arr = Exam_HMM.Best_States()
print(Exam_HMM.score)`
Can you give dataset to your code and show the best states for example?

**Comment by manitadayon** [comment]
I looked at the data. There are two options: 1- Unsupervised HMM, you can use Auto-HMM and ignore the label column and please also ignore the time stamp. 2- Use the label column (still ignore the timestamp) and use the supervised packages such as this in MATLAB which is very easy https://www.mathworks.com/help/stats/hidden-markov-models-hmm.html

You can also try this:

from Hidden_Markov_Model import *
from Hidden_Markov_Model.HMM import Supervised_HMM
import time

Start=time.time()
Train_ratio=0.9
Cov_Type='diag'
Max_state=7
Max_mixture=4
Iter=100
Feat=1
N=3
T=10000
flag=1
Path = 'Path to CSV'
Data=pd.read_csv(Path)
Data2 = Data.iloc[:,2:5]
Data2 = Data2.T
Exam_HMM=Supervised_HMM(Train_ratio,Cov_Type,Max_state,Max_mixture,Iter,Feat,N,T,Data2,flag)
Exam_HMM.Best_States()
END=time.time()
print('Total Time Takes in seconds',END-Start)

You can also use hmmlearn and feed the observation separately instead of using Auto-HMM.

**Comment by moryekram** [comment]
Hi again. Excuse me. 

This is my result with params and using "hmm_classifier" library:

Data=pd.read_csv(Path, header=0)
x = Data.iloc[:,3:6].to_numpy()
x = x.reshape(64645,1,3)
y = Data.iloc[:,6:7].to_numpy()
y = y.reshape(-1)
model = HMM_classifier(hmm.GaussianHMM())
model.fit(x,y)
pred = model.predict_proba([[-0.352388077,0.502128386,0.390059343]])

library: hmm_classifier -> https://github.com/elmihailol/HMM-classifier
result: error: rows of transmat_ must sum to 1.0 (got {0})

Can you show me how to use hmmlearn step by step to pass 70% data for train and 30% for test and get probability for each class and also get accuracy but with my data? I am so confused and my time is running out, Thanks

**Comment by moryekram** [comment]
This is my result whit params and using "hmm_classifier" library:

Data=pd.read_csv(Path, header=0)
x = Data.iloc[:,3:6].to_numpy()
x = x.reshape(64645,1,3)
y = Data.iloc[:,6:7].to_numpy()
y = y.reshape(-1)
# param=set(x.ravel())
model = HMM_classifier(hmm.GaussianHMM())
model.fit(x,y)
pred = model.predict_proba([[-0.352388077,0.502128386,0.390059343]])

library: hmm_classifier -> https://github.com/elmihailol/HMM-classifier
result: error: rows of transmat_ must sum to 1.0 (got {0})
